File: Recursion/Lecture 07/CodeStudio/Power Set.py
def pwset(v):
    def helper(v, i, subSet, ans):
        
        # Our base case will hit when our i pointer reaches the end of given array v[]
        if i == len(v):
            ans.append(subSet.copy())
            return
        
        # we pick ith Element
        subSet.append(v[i])
        
        # We ask recursion to do rest of the task
        helper(v, i+1, subSet, ans)
        
        # Backtrack and Undo the change we have done
        subSet.pop()
        
        # We don't pick the ith element
        helper(v, i+1, subSet, ans)
    
    # varibles in pwset() scope
    ans = []
    subSet = []
    # Helper will store the ans in ans list        
    helper(v, 0, subSet, ans)
    return ans

File: Recursion/Lecture 07/CodeStudio/Power Set.cpp
void helper(vector<int> &v, int i, vector<int> &subSet, vector<vector<int>> &ans)
{
    // base condition
    if (i == v.size())
    {
        ans.push_back(subSet);
        return;
    }

    // pick the ith element
    subSet.push_back(v[i]);
    helper(v, i + 1, subSet, ans);

    // backtrack & undo the changes that were made
    subSet.pop_back();

    // skip the ith element
    helper(v, i + 1, subSet, ans);
}

vector<vector<int>> pwset(vector<int> v)
{
    vector<vector<int>> ans;
    vector<int> subSet;
    helper(v, 0, subSet, ans);
    return ans;
}

/*
    time complexity : O(2^n)
    space complexity : O(n)
*/